controller/src/main/java/io/pravega/controller/store/stream/StreamConfigWithVersion.java

```diff
@@ -0,0 +1,28 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.store.stream;
+
+import io.pravega.client.stream.StreamConfiguration;
+import lombok.Data;
+
+import java.io.Serializable;
+
+@Data
+public class StreamConfigWithVersion implements Serializable {
+
+    private final StreamConfiguration configuration;
+
+    private final int version;
+
+    public static StreamConfigWithVersion generateNext(final StreamConfigWithVersion previous,
```

Comment:
The purpose version serves is as follows:
If an update operation is rerun (because of a failover) we do not want to override a newer configuration with an older one. Putting a version along with the configuration data prevents us from doing this. 
So before updating configuration, we check if the version allows us to do so. 

Now to answer your question: 
> As such, doesn't really guarantee that versions are unique or even monotonically increasing
Yes, this api does not guarantee that. However, logically this version is monotonically increasing because when we actually update this data in the store, since our store supports CAS, any older version update attempt fails.

Reply:
From your description, it looks like we are using two versions: the one added here and the one of the store. Isn't it redundant?

Reply:
Its not actually. Both have different semantic purpose. 
The store's version is something that the store is aware of. Rest of the layers are not aware of that version. Store abstracts that. And the store uses a generic type of version whose versioned comparison is store's responsibility. 
The store's version is for compare and swap semantics. Do a get-and-set on an object and if there is a conflict because of concurrent updates, fail the one that used the older version. 

This version we have here is used to ensure that the configuration is not re-updated upon failover recovery because we replay and reprocess and older event. 
 
Imagine following scenario: two updateConfig requests U1 and U2 followed by a failover. 
- requeststream: ...  U1, U2 ... 
- checkpoint before U1

Now a failover happens. and we replay U1. We dont want to reprocess U1 at all. For that we need to know that processing has already moved ahead.

yes, we could have reused the zk store's version for this purpose. But our store interface uses a generic for the version.

Reply:
Perhaps we shouldn't it here, but I'm wondering if it is worth re-evaluating eventually the version exposed by the interface so that generic facility can be reused by upper layers. It simplifies the reasoning if we have a single version.

If you agree, please create an issue.